### PR TITLE
feat: capactity to discard meshes and use its diffuseColor, effect damage && create a batch exclusively for Characters

### PR DESCRIPTION
**Pull request by PaablooCH**
Things added:

- In order to create the effect I add the capacity to choose if we want to discard (not render, for the people outside graphics :)) or choose an additive color for a mesh. You can test the discard implementation through interface.
- Too make it less cost that change I create a new category for batches to store Characters, who are the ones that must use the effect. Characters are the ones who uses the tag `Player` or `Enemy` and his children, when you modify a tag if now is a character or was a character, the gameObject and his children will be moved to the correct batch.
- Create a script to manage the effects.

If you want to test the effect go to gameplay scene and let you be hitted

Other things:

- Clear objectIndexes in batch before fill it again.
- Solve warings.

### CODE REVIEW

File: Source/DataModels/Batch/BatchManager.cpp [hunk 1]
```diff
@@ -17,6 +22,25 @@ BatchManager::~BatchManager()
...
+
+void BatchManager::SwapBatchParentAndChildren(GameObject* go)
+{
+	ComponentMeshRenderer* component = go->GetComponentInternal<ComponentMeshRenderer>();
+	if (component)
+	{
+		component->GetBatch()->DeleteComponent(component);
+		AddComponent(component);
+	}
+	for (auto child : go->GetChildren())
```
Comment (markmeliaa) [inline Source/DataModels/Batch/BatchManager.cpp:38]: Couldn't this `auto` be a `GameObject*`?

File: Source/DataModels/Batch/BatchManager.cpp [hunk 2]
```diff
@@ -43,10 +67,27 @@ void BatchManager::AddComponent(ComponentMeshRenderer* newComponent)
...
 }
 
+void BatchManager::SearchAndSwapBatchCharacter(GameObject* parent)
+{
+	if (parent->GetTag() == "Player" || parent->GetTag() == "Enemy")
+	{
+		SwapBatchParentAndChildren(parent);
+		return;
+	}
+	for (auto child : parent->GetChildren())
```
Comment (markmeliaa) [inline Source/DataModels/Batch/BatchManager.cpp:77]: Same with this `auto`.

File: Source/Modules/ModuleRender.cpp
Comment (markmeliaa) [inline Source/Modules/ModuleRender.cpp:1]: Lines 757, 803 and 894 raise a warning, could you take a look at them please? Not really related in this pr.

**Review by markmeliaa — CHANGES_REQUESTED** [review]
Nice job, tried it and works correctly, my only complain is that when the mesh turns red after a hit it is for a really brief period of time, maybe we could extend a little bit that time so it is easier for the player to see the effect?

P.S. Didn't check the shaders because my knowledge about them is practically non-existent.

**Comment by PaablooCH** [comment]
> maybe we could extend a little bit that time so it is easier for the player to see the effect?

It can be tweaked with the values `TIME_BETWEEN_EFFECTS` and `MAX_TIME_EFFECT_DURATION` so is not a big problem, if you want you can also add more transitions like `color-discard-color` or `discard-color` instead of what is now `color-discard`.

File: Source/DataModels/Batch/BatchManager.cpp (hunk 1 above)
Comment (PaablooCH) [reply Source/DataModels/Batch/BatchManager.cpp:38]: i use auto to avoid writting each time GameObject*, it doesn't affect the performance.

**Review by markmeliaa — APPROVED** [review]

**Comment by UriMurcia** [comment]
Hi! I've haven't time to see all the code and test it, but from the part of gameplay I think it would be interesting that the functions that change the color of the mesh wouldn't depend from HealthSystem script, we may want to use these functionality for other purposes, for example, when the drone is seeking the player to explode.
I'll take a look to the other things when I have time! :)

**Review by markmeliaa — CHANGES_REQUESTED** [review]
*solved*

**Comment by MacXxs** [comment]
We need to first approve PR #647 before this one in order to adapt the percentage selection from texturing into the tiling changes done in this PR (that maybe shouldn't been done in this one actually in order to not have any trouble while merging with developments from other PRs)

File: Game/Scripts/MeshEffect.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	for (auto mesh : meshes)
+	{
+		mesh->SetEffectColor(float3(0.f, 0.f, 0.f));
+		mesh->SetDiscard(true);
+	}
+}
+
+void MeshEffect::EffectColor(float3 color)
+{
+	for (auto mesh : meshes)
```
Comment (markmeliaa) [inline Game/Scripts/MeshEffect.cpp:90]:
```suggestion
	for (ComponentMeshRenderer* mesh : meshes)
```

File: Game/Scripts/MeshEffect.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	if (effectTime > maxTime)
+	{
+		activateEffect = false;
+		ClearEffect();
+	}
+}
+
+void MeshEffect::EffectDiscard()
+{
+	for (auto mesh : meshes)
```
Comment (markmeliaa) [inline Game/Scripts/MeshEffect.cpp:81]:
```suggestion
	for (ComponentMeshRenderer* mesh : meshes)
```

File: Game/Scripts/MeshEffect.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+{
+	this->timeBetween = timeBetween;
+	this->maxTime = maxTime;
+	activateEffect = true;
+	effectTime = 0.f;
+}
+
+void MeshEffect::ClearEffect()
+{
+	for (auto mesh : meshes)
```
Comment (markmeliaa) [inline Game/Scripts/MeshEffect.cpp:49]:
```suggestion
	for (ComponentMeshRenderer* mesh : meshes)
```

File: Source/DataModels/Scene/Scene.cpp [hunk 6]
```diff
@@ -127,8 +127,13 @@ std::vector<GameObject*> Scene::ObtainObjectsInFrustum(const math::Frustum* frus
 	}
 
 #ifdef ENGINE
-	CalculateNonStaticObjectsInFrustum(frustum, App->GetModule<ModuleScene>()->GetSelectedGameObject(), 
-									   objectsInFrustum);
+	GameObject* selected = App->GetModule<ModuleScene>()->GetSelectedGameObject();
+	CalculateNonStaticObjectsInFrustum(frustum, selected, objectsInFrustum);
+
+	for (auto childSelected : selected->GetChildren())
```
Comment (markmeliaa) [inline Source/DataModels/Scene/Scene.cpp:133]:
```suggestion
	for (GameObject* childSelected : selected->GetChildren())
```

**Review by markmeliaa — COMMENTED** [review]
I'm failing to see the effect happening now, is the damage effect not red anymore or is it so fast that I can't even percieve it?

File: Source/DataModels/Scene/Scene.cpp (hunk 6 above)
Comment (markmeliaa) [reply Source/DataModels/Scene/Scene.cpp:133]: Pablo ignora mis requests.

**Review by markmeliaa — APPROVED** [review]
LGTM now! (La proxima vez acosa a otro, grax).

**Comment by MacXxs** [comment]
Level1 throws error when pressing play:

image

Try merging again with Graphics-main, once it works well i'll approve the PR.

**Comment by PaablooCH** [comment]
> Level1 throws error when pressing play:
> 
> image
> 
> Try merging again with Graphics-main, once it works well i'll approve the PR.

Lvl1 doen't have the script Mesh Effect, remember we can not touch lvls ;) to avoid problems with level team, only is implemented in gameplay scene and prefabs

**Comment by MacXxs** [comment]
> > Level1 throws error when pressing play:
> > image
> > Try merging again with Graphics-main, once it works well i'll approve the PR.
> 
> Lvl1 doen't have the script Mesh Effect, remember we can not touch lvls ;) to avoid problems with level team, only is implemented in gameplay scene and prefabs

In the Graphics-main branch level1 is playable, and this PR is a branch from Graphics-main, so the same level should be playable.

**Review by MacXxs — APPROVED** [review]

**Pull request merged**